src/org/mozilla/javascript/NativeConsole.java

```diff
@@ -0,0 +1,437 @@
+            return;
+        }
+
+        StringBuilder msg = new StringBuilder("Assertion failed:");
+        if (args != null && args.length > 1) {
+            for (int i = 1; i < args.length; ++i) {
+                msg.append(" ").append(ScriptRuntime.toString(args[i]));
+            }
+        } else {
+            msg.append(" console.assert");
+        }
+
+        print(cx, scope, Level.ERROR, msg.toString());
+    }
+
+    private void count(Context cx, Scriptable scope, Object[] args) {
+        String label = args.length > 0 ? ScriptRuntime.toString(args[0]) : DEFAULT_LABEL;
+        int count = counters.computeIfAbsent(label, l -> new AtomicInteger(0)).incrementAndGet();
+        print(cx, scope, Level.INFO, label + ": " + count);
+    }
+
+    private void countReset(Context cx, Scriptable scope, Object[] args) {
+        String label = args.length > 0 ? ScriptRuntime.toString(args[0]) : DEFAULT_LABEL;
+        AtomicInteger counter = counters.remove(label);
+        if (counter == null) {
+            print(cx, scope, Level.WARN, "Count for '" + label + "' does not exist.");
+        }
+    }
+
+    private void time(Context cx, Scriptable scope, Object[] args) {
+        String label = args.length > 0 ? ScriptRuntime.toString(args[0]) : DEFAULT_LABEL;
+        Long start = timers.get(label);
+        if (start != null) {
+            print(cx, scope, Level.WARN, "Timer '" + label + "' already exists.");
+            return;
+        }
+        timers.put(label, System.nanoTime());
+    }
+
+    private void timeEnd(Context cx, Scriptable scope, Object[] args) {
+        String label = args.length > 0 ? ScriptRuntime.toString(args[0]) : DEFAULT_LABEL;
+        Long start = timers.remove(label);
+        if (start == null) {
+            print(cx, scope, Level.WARN, "Timer '" + label + "' does not exist.");
+            return;
+        }
+        print(cx, scope, Level.INFO, label + ": " + nano2Milli(System.nanoTime() - start) + "ms");
+    }
+
+    private void timeLog(Context cx, Scriptable scope, Object[] args) {
+        String label = args.length > 0 ? ScriptRuntime.toString(args[0]) : DEFAULT_LABEL;
+        Long start = timers.get(label);
+        if (start == null) {
+            print(cx, scope, Level.WARN, "Timer '" + label + "' does not exist.");
+            return;
+        }
+        StringBuilder msg =
+                new StringBuilder(label + ": " + nano2Milli(System.nanoTime() - start) + "ms");
+
+        if (args.length > 1) {
+            for (int i = 1; i < args.length; i++) {
+                msg.append(" ").append(ScriptRuntime.toString(args[i]));
+            }
+        }
+        print(cx, scope, Level.INFO, msg.toString());
+    }
+
+    private double nano2Milli(Long nano) {
+        return nano / 1000000D;
+    }
+
+    // #string_id_map#
+
+    @Override
+    protected int findPrototypeId(String s) {
+        int id;
+        // #generated# Last update: 2021-11-10 14:43:23 CST
```

Comment:
ref. #848, #914
The latest code now uses String in Switch Case, and `idswitch` has been removed.

Reply:
:+1: [ae1b6a9](https://github.com/mozilla/rhino/pull/1103/commits/ae1b6a9981bfecc668dda5d67fd1bbc10a74d659)

Reply:
Done.